### ISSUE TITLE
server profile update in-place

### ISSUE THREAD

**Issue by rismoney** [HewlettPackard/terraform-provider-oneview#295]
Having a strange issue-

some background
Creating a server profile based on a template.  The template has bios options in it.
The server profile has local storage and connection settings in it.

When I create TF apply the server profile, it creates exactly as expected. Excellent!
However:
If I rerun TF plan/apply, TF tries to improperly "change" the resource, specifically the bios settings are planned to -> null and  connection settings ->null and then connections wants to do a delete/add operation.

I am wondering if the state file is problematic here on this?
Let me know if you need sample code.

@AsisBagga Thank you so much for all your help getting this going.

**Comment by rismoney** [comment]
Here is sample output truncated to show what i mean

```
  # module.foo.oneview_server_profile.default will be updated in-place
  ~ resource "oneview_server_profile" "default" {
        affinity              = "Bay"
        boot_order            = []
        enclosure_group       = "EG_01"
        hardware_name         = "EG01_FRAME01, bay 1"
        hardware_uri          = "/rest/server-hardware/39313738-3034-584D-5131-303230483230"
        hide_unused_flex_nics = true
        id                    = "SP_EG01_FRAME01_B01"
        ilo_ip                = "172.20.172.17"
        mac_type              = "Physical"
        name                  = "SP_EG01_FRAME01_B01"
        serial_number         = "MXQ1020H20"
        serial_number_type    = "Physical"
        server_hardware_type  = "SY 480 Gen10 1"
        template              = "foo"
        type                  = "ServerProfileV12"
        uri                   = "/rest/server-profiles/87a1ad26-a7d5-42b7-92df-9c4eaba40502"
        wwn_type              = "Physical"

      - bios_option {
          - consistency_state = "Consistent" -> null
          - manage_bios       = true -> null
          - reapply_state     = "NotApplying" -> null

          - overridden_settings {
              - id    = "CollabPowerControl" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "EnergyEfficientTurbo" -> null
              - value = "Disabled" -> null
            }

        }

        boot_mode {
            manage_mode     = true
            mode            = "UEFI"
            pxe_boot_policy = "Auto"
            secure_boot     = "unmanaged"
        }

      - connection_settings {
          - manage_connections = false -> null

          - connections {
              - allocated_mbps    = 0 -> null
              - allocated_vfs     = 0 -> null
              - function_type     = "Ethernet" -> null
              - id                = 0 -> null
              - interconnect_port = 0 -> null
              - isolated_trunk    = false -> null
              - managed           = false -> null
              - maximum_mbps      = 0 -> null
              - name              = "G1_Network_1" -> null
              - network_uri       = "/rest/ethernet-networks/44155074-0d29-4dcd-ba76-ec095130fc36" -> null
              - port_id           = "Mezz 3:1-a" -> null
              - requested_mbps    = "1000" -> null
            }
          - connections {
              - allocated_mbps    = 0 -> null
              - allocated_vfs     = 0 -> null
              - function_type     = "Ethernet" -> null
              - id                = 0 -> null
              - interconnect_port = 0 -> null
              - isolated_trunk    = false -> null
              - managed           = false -> null
              - maximum_mbps      = 0 -> null
              - name              = "G1_Network_2" -> null
              - network_uri       = "null" -> null
              - port_id           = "Mezz 3:2-a" -> null
              - requested_mbps    = "0" -> null
            }
        }
      + connection_settings {
          + connections {
              + function_type  = "Ethernet"
              + id             = 1
              + name           = "Network_1"
              + network_uri    = "/rest/ethernet-networks/44155074-0d29-4dcd-ba76-ec095130fc36"
              + port_id        = "Mezz 3:1-a"
              + requested_mbps = "1000"

              + boot {
                  + priority = "Primary"
                }
            }
          + connections {
              + function_type  = "Ethernet"
              + id             = 2
              + name           = "Network_2"
              + port_id        = "Mezz 3:2-a"
              + requested_mbps = "0"

              + boot {
                  + priority = "Secondary"
                }
            }
        }
    }
```
Plan: 0 to add, 1 to change, 0 to destroy.

**Comment by rismoney** [comment]
my temporary fix was to use a lifecycle ignore changes on connection_settings, and bios_option , but ideally will want this work as intended.

**Comment by AsisBagga** [comment]
@rismoney Thanks Rich, we are aware of this issue, and we have implemented the fix, this should be addressed in our coming release.

**Comment by rismoney** [comment]
If necessary, I can test a branch when it's ready

**Comment by AsisBagga** [comment]
> If necessary, I can test a branch when it's ready

sure you may clone and build the binary for testing, and let us know.

**Comment by rismoney** [comment]
I tested the #296 branch.


Connection Settings - This looks like mac_type,managed, ethernet_boot_type are trying to be => null but definitely improved over above.

For the bios_option same as above output.



```
~ connection_settings {
            manage_connections = false

          ~ connections {
                allocated_mbps    = 0
                allocated_vfs     = 0
                function_type     = "Ethernet"
                id                = 1
                interconnect_port = 0
                isolated_trunk    = false
              - mac_type          = "Physical" -> null
              - managed           = true -> null
                maximum_mbps      = 0
                name              = "G1_Network_1"
                network_uri       = "/rest/ethernet-networks/44155074-0d29-4dcd-ba76-ec095130fc36"
                port_id           = "Mezz 3:1-a"
                requested_mbps    = "1000"

              ~ boot {
                    boot_vlan_id       = 0
                  - ethernet_boot_type = "PXE" -> null
                    priority           = "Primary"
                }
            }
          ~ connections {
                allocated_mbps    = 0
                allocated_vfs     = 0
                function_type     = "Ethernet"
                id                = 2
                interconnect_port = 0
                isolated_trunk    = false
              - mac_type          = "Physical" -> null
              - managed           = true -> null
                maximum_mbps      = 0
                name              = "G1_Network_2"
                port_id           = "Mezz 3:2-a"
                requested_mbps    = "0"

              ~ boot {
                    boot_vlan_id       = 0
                  - ethernet_boot_type = "PXE" -> null
                    priority           = "Secondary"
                }
            }
        }
    }
```

**Comment by rismoney** [comment]
not sure this is related @AsisBagga 


I am now getting using a module I have deployed 42 times within this enclosure group, but now I am struggling on 6 servers with this error. 4 within this frame built fine.  bays 1/2 are empty only in this frame, not sure if that changes anything.

Trying to figure out why TF is failing.


```
2021/04/19 12:10:03 [DEBUG] module.abc.oneview_server_profile.default: applying the planned Create change
w_server_profi2021/04/19 12:10:03 [DEBUG] moduleabc.oneview_server_profile.default: apply errored, but we're indicating that via the Error pointer rather than returning it: Provider produced inconsistent result after apply: When applying changes to module.abconeview_server_profile.default, provider "registry.terraform.io/-/oneview" produced an unexpected new value for was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
le2021/04/19 12:10:03 [ERROR] module.abc: eval: *terraform.EvalApplyPost, err: Provider produced inconsistent result after apply: When applying changes to module.abc.oneview_server_profile.default, provider "registry.terraform.io/-/oneview" produced an unexpected new value for was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
2021/04/19 12:10:03 [ERROR] module.abc: eval: *terraform.EvalSequence, err: Provider produced inconsistent result after apply: When applying changes to module.abc.oneview_server_profile.default, provider "registry.terraform.io/-/oneview" produced an unexpected new value for was present, but now absent.
```

**Comment by rismoney** [comment]
I seem to be hitting a bug also where sometimes it seems like it's misidentifying a profile to modify.  It's not every TF run, it's seemingly random:

```
# module.MIAQMS05.oneview_server_profile.default will be updated in-place
  ~ resource "oneview_server_profile" "default" {
        affinity              = "Bay"
        enclosure_group       = "EG_01"
        hardware_name         = "EG01_FRAME03, bay 1"
        hardware_uri          = "/rest/server-hardware/39313738-3034-584D-5131-303230483250"
        hide_unused_flex_nics = true
        id                    = "SP_EG01_FRAME03_B01"
        ilo_ip                = "172.20.172.29"
        mac_type              = "Physical"
      ~ name                  = "SP_EG01_FRAME01_B04" -> "SP_EG01_FRAME03_B01"
        serial_number         = "64546456546"
        serial_number_type    = "Physical"
...
```
If I rerun TF this update in place seems to go away. And then reappear on subequent runs. 
Upon first impression it looks like the hardware URI lookups are wrong or are getting cached across resources.

**Comment by rismoney** [comment]
deleted some unrelated msgs - had a typo.  sorry for confusion.

**Comment by rismoney** [comment]
i will also add that the PR #296 in it's current form isn't quite right yet.  It was buggy on a few things, tries to make the boot stanza args fully null.  I tried then putting an ignore_changes on boot too, but then  that got weird, since technically the boot isn't specified in the profile (defaults?)  and weird hardware uri lookups started happening with strange changes being suggested intermittently.

**Comment by rismoney** [comment]
i stand corrected- the hardware lookup thing is random, and occurs on master branch - kinda dangerous bug.  Here's an example (doesn't happen everytime)


~ resource "oneview_server_profile" "default" {
        affinity              = "Bay"
        boot_order            = []
        enclosure_group       = "EG_01"
        hardware_name         = "EG01_FRAME01, bay 2"
        hardware_uri          = "/rest/server-hardware/39313738-3034-584D-5131-303230483250"
        hide_unused_flex_nics = true
        id                    = "SP_EG01_FRAME01_B02"
        ilo_ip                = "172.20.172.2"
        mac_type              = "Physical"
      ~ name                  = "SP_EG01_FRAME01_B04" -> "SP_EG01_FRAME01_B02"
        serial_number         = "abc123"
        serial_number_type    = "Physical"
        server_hardware_type  = "SY 480 Gen10 1"
        template              = "SPT_EG01_Desktops_Gen10_1"
        type                  = "ServerProfileV12"
        uri                   = "/rest/server-profiles/030d5c76-8796-462b-a801-311764fa349c"
        wwn_type              = "Physical"

**Comment by AsisBagga** [comment]
@rismoney Hi Rich, can you explain more about hardware look up?

And for mac_type,managed, ethernet_boot_type and bios_option attributes in connections is going null maybe because you haven't provided them in your example file, would you mind sharing you example file with me?
I tried and it works fine for me...

**Comment by rismoney** [comment]
hardware lookup weirdness: when i was using the #296 branch it seemed as those it didn't always resolve the correct profile to modify.   in my message right before this, i ran terraform plan, and it should always be a noop, as SP_EG01_FRAME02 is correct for frame 1 bay 2.  However the provider somehow randomly thinks (doesn't happen deterministically every plan) that SP_EG01_FRAME04 is the current name, and that it wants to change it.   The hardware uri ending in 3250 is associated with Bay 4.
So somehow the looks aren't happening reliably to match against state or something like that.   Again, I could re-run plan, and this then may not show up.   And again - it could.  

In this directory I am currently managing 40 profiles via modules.  I am not sure if I am hitting some other weirdness, but the other 38 seem fine as far as i can tell.  But there is literally no difference between them, aside from names in most cases.

As far as the resource - sure:
What is interesting about the above output is that for example mac_type is under connection(s).  I think that is and wwn is per profile setting not per connnection.

For the bios settings, I don't have any defined in the server profile.  They are in the server profile template, so it looks like it is munging the 2, and thinks because I don't have them defined is server profile, that they should be removed.



```
locals { 
  netmappings = jsondecode(data.local_file.mappings.content)
  frameparity = tonumber(substr(var.profile_name,6,1))
  enclosuregroup_no = substr(var.profile_name,5,2)
  frame = substr(var.profile_name,8,7)
  bay = tonumber(substr(var.profile_name,17,2)) 
}

data "local_file" "mappings" {
  filename = (local.frameparity % 2 == 1 ) ? "${path.module}/oddframes/${var.environment}-${var.server_function}.json" : "${path.module}/evenframes/${var.environment}-${var.server_function}.json"
}

data "oneview_ethernet_network" "network" {
  count = length(local.netmappings.connections)
  name = lookup(local.netmappings.connections[count.index],"networkUri")
}

resource "oneview_server_profile" "default" {
  name = var.profile_name
  template = var.profile_template
  hardware_name = "EG${local.enclosuregroup_no}_${local.frame}, bay ${local.bay}"
  wwn_type = "Physical"
  mac_type = "Physical"
  enclosure_group = "EG_${local.enclosuregroup_no}"
  server_hardware_type = var.server_hardware_type
  serial_number_type = "Physical"
  type = "ServerProfileV12"
  boot_mode {
    manage_mode = true
    mode = "UEFI"
    pxe_boot_policy = "Auto"  
    secure_boot="unmanaged"
  }
  
  connection_settings {
  
    dynamic "connections" {
      for_each = data.oneview_ethernet_network.network
      content {
        network_uri = connections.value.uri
        name = lookup(local.netmappings.connections[connections.key],"name")
        id = lookup(local.netmappings.connections[connections.key],"id")      
        function_type = lookup(local.netmappings.connections[connections.key],"function_type","Ethernet")
        port_id = lookup(local.netmappings.connections[connections.key],"portId")
        requested_mbps = lookup(local.netmappings.connections[connections.key],"requestedMbps")
        boot  {
          priority           = contains(keys(local.netmappings.connections[connections.key]), "boot") ? lookup(local.netmappings.connections[connections.key]["boot"],"priority","")  : null
          ethernet_boot_type = contains(keys(local.netmappings.connections[connections.key]), "boot") ? lookup(local.netmappings.connections[connections.key]["boot"],"ethernet_boot_type","") : null
        }
      }
    }
  } 
```

**Comment by AsisBagga** [comment]
The reason terraform is looking to change the Server Profile Name has nothing to do with which server hardware which is assigned rather it will check the local resource(server profile) Id(name) and matches the one within oneview and updates if differ. For me the best way to look the "hardware URI assignment" is by matching the Server Profile URI with server hardware URI. 

In your terraform show it looks like profile uri is  "/rest/server-profiles/030d5c76-8796-462b-a801-311764fa349c"
and hardware_uri = "/rest/server-hardware/39313738-3034-584D-5131-303230483250"
and terraform local id = "SP_EG01_FRAME01_B02"

so maybe due to multiple different iterations on same ID, the terraform state file is messed up, What i would suggest is either you can do an import and then update to check for any anomalies.

PS: For server profile update you have to mention update_type = "put" inorder to make your update work, I hope youre aware of that.
PS: You may open a new issue for hardware lookup next time if you encounter any anomaly.

**Comment by AsisBagga** [comment]
> hardware lookup weirdness: when i was using the #296 branch it seemed as those it didn't always resolve the correct profile to modify. in my message right before this, i ran terraform plan, and it should always be a noop, as SP_EG01_FRAME02 is correct for frame 1 bay 2. However the provider somehow randomly thinks (doesn't happen deterministically every plan) that SP_EG01_FRAME04 is the current name, and that it wants to change it. The hardware uri ending in 3250 is associated with Bay 4.
> So somehow the lookups aren't happening reliably to match against state or something like that. Again, I could re-run plan, and this then may not show up. And again - it could.
> 
> In this directory I am currently managing 40 profiles via modules. I am not sure if I am hitting some other weirdness, but the other 38 seem fine as far as i can tell. But there is literally no difference between them, aside from names in most cases.
> 
> As far as the resource - sure:
> What is interesting about the above output is that for example mac_type is under connection(s). I think that is and wwn is per profile setting not per connnection.
> 
> For the bios settings, I don't have any defined in the server profile. They are in the server profile template, so it looks like it is munging the 2, and thinks because I don't have them defined is server profile, that they should be removed.
> 
> ```
> locals { 
>   netmappings = jsondecode(data.local_file.mappings.content)
>   frameparity = tonumber(substr(var.profile_name,6,1))
>   enclosuregroup_no = substr(var.profile_name,5,2)
>   frame = substr(var.profile_name,8,7)
>   bay = tonumber(substr(var.profile_name,17,2)) 
> }
> 
> data "local_file" "mappings" {
>   filename = (local.frameparity % 2 == 1 ) ? "${path.module}/oddframes/${var.environment}-${var.server_function}.json" : "${path.module}/evenframes/${var.environment}-${var.server_function}.json"
> }
> 
> data "oneview_ethernet_network" "network" {
>   count = length(local.netmappings.connections)
>   name = lookup(local.netmappings.connections[count.index],"networkUri")
> }
> 
> resource "oneview_server_profile" "default" {
>   name = var.profile_name
>   template = var.profile_template
>   hardware_name = "EG${local.enclosuregroup_no}_${local.frame}, bay ${local.bay}"
>   wwn_type = "Physical"
>   mac_type = "Physical"
>   enclosure_group = "EG_${local.enclosuregroup_no}"
>   server_hardware_type = var.server_hardware_type
>   serial_number_type = "Physical"
>   type = "ServerProfileV12"
>   boot_mode {
>     manage_mode = true
>     mode = "UEFI"
>     pxe_boot_policy = "Auto"  
>     secure_boot="unmanaged"
>   }
>   
>   connection_settings {
>   
>     dynamic "connections" {
>       for_each = data.oneview_ethernet_network.network
>       content {
>         network_uri = connections.value.uri
>         name = lookup(local.netmappings.connections[connections.key],"name")
>         id = lookup(local.netmappings.connections[connections.key],"id")      
>         function_type = lookup(local.netmappings.connections[connections.key],"function_type","Ethernet")
>         port_id = lookup(local.netmappings.connections[connections.key],"portId")
>         requested_mbps = lookup(local.netmappings.connections[connections.key],"requestedMbps")
>         boot  {
>           priority           = contains(keys(local.netmappings.connections[connections.key]), "boot") ? lookup(local.netmappings.connections[connections.key]["boot"],"priority","")  : null
>           ethernet_boot_type = contains(keys(local.netmappings.connections[connections.key]), "boot") ? lookup(local.netmappings.connections[connections.key]["boot"],"ethernet_boot_type","") : null
>         }
>       }
>     }
>   } 
> ```

The reason terraform is looking to change the Server Profile Name has nothing to do with which server hardware which is assigned rather it will check the local resource(server profile) Id(name) and matches the one within oneview and updates if differ. For me the best way to look the "hardware URI assignment" is by matching the Server Profile URI with server hardware URI.

In your terraform show it looks like profile uri is "/rest/server-profiles/030d5c76-8796-462b-a801-311764fa349c"
and hardware_uri = "/rest/server-hardware/39313738-3034-584D-5131-303230483250"
and terraform local id = "SP_EG01_FRAME01_B02"

so maybe due to multiple different iterations on same ID, the terraform state file is messed up, What i would suggest is either you can do an import and then update to check for any anomalies.

PS: For server profile update you have to mention update_type = "put" inorder to make your update work, I hope youre aware of that.
PS: You may open a new issue for hardware lookup next time if you encounter any anomaly.

**Comment by rismoney** [comment]
Like I said the weirdness with this newer branch, was intermittent.  Will continue to test as I see it was merged.

I am hoping that update_type=put won't be required in the future, as terraform should know the difference between apply changes that are "new resources" versus updating existing ones.  This argument seems unintuitive. (Like how am I supposed to know (or care) what state this option should be in)

This seems silly that the end user should need to know the mechanics of the restapi call as it's an implementation detail.   I didn't see that in the example either, and the doc only mentions put or patch.

Thanks for all your help!  PRovider getting better everyday!

**Comment by rismoney** [comment]
Ok, started testing new provider.

So what I think happened here, is that the state file has the bios and local storage settings in it.  My belief is that shouldn't have been in there as they were part of the server profile template.  so now the new version isn't putting it in there, and that is why it's trying to remove it.

I am going to do some state file surgery and see what's up.

**Comment by rismoney** [comment]
ok-  the provider in the master branch no longer will work in my configuration.

To be clear I have a server profile template that contains bios settings, and local storage.
I then have a server profile based on that references that template
I have my network configuration within the profile.

I hope that clears up the confusion.

I get the same results on importing an existing server profile template versus an existing resource:
To summarize, it looks like it's munging the server profile settings with the inherited settings from server profile template.
The server profile resource doesn't have local storage or bios, therefore when you plan on that resource, it looks like it's trying to remove those settings.

Then there are other settings, that are curiously not correct.  I enclosed a sample of what i see



This is an example of a newly imported server profile resource.
```
# module.pkini6.oneview_server_profile.default will be updated in-place
  ~ resource "oneview_server_profile" "default" {
        affinity              = "Bay"
        enclosure_group       = "EG_01"
      + hardware_name         = "EG01_FRAME03, bay 5"
        hardware_uri          = "/rest/server-hardware/39313738-3034-584D-5131-303230483135"
        hide_unused_flex_nics = true
        id                    = "SP_EG01_FRAME03_B05"
        ilo_ip                = "172.20.172.30"
        mac_type              = "Physical"
        name                  = "SP_EG01_FRAME03_B05"
        serial_number         = "MXQ1020H15"
        serial_number_type    = "Physical"
        server_hardware_type  = "SY 480 Gen10 1"
      + template              = "SPT_EG01_Desktops_Gen10_1"
        type                  = "ServerProfileV12"
        uri                   = "/rest/server-profiles/db366936-a971-427d-b383-0d2e2f8703ec"
        wwn_type              = "Physical"

      - bios_option {
          - consistency_state = "Consistent" -> null
          - manage_bios       = true -> null
          - reapply_state     = "NotApplying" -> null

          - overridden_settings {
              - id    = "MinProcIdlePower" -> null
              - value = "NoCStates" -> null
            }
          - overridden_settings {
              - id    = "WorkloadProfile" -> null
              - value = "Custom" -> null
            }
          - overridden_settings {
              - id    = "IntelUpiPowerManagement" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "ProcHyperthreading" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "UncoreFreqScaling" -> null
              - value = "Maximum" -> null
            }
          - overridden_settings {
              - id    = "EnergyPerfBias" -> null
              - value = "MaxPerf" -> null
            }
          - overridden_settings {
              - id    = "MinProcIdlePkgState" -> null
              - value = "NoState" -> null
            }
          - overridden_settings {
              - id    = "MemPatrolScrubbing" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "ProcTurbo" -> null
              - value = "Enabled" -> null
            }
          - overridden_settings {
              - id    = "TimeZone" -> null
              - value = "UtcM5" -> null
            }
          - overridden_settings {
              - id    = "ProcX2Apic" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "PowerRegulator" -> null
              - value = "StaticHighPerf" -> null
            }
          - overridden_settings {
              - id    = "EnergyEfficientTurbo" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "NumaGroupSizeOpt" -> null
              - value = "Clustered" -> null
            }
          - overridden_settings {
              - id    = "MemRefreshRate" -> null
              - value = "Refreshx1" -> null
            }
          - overridden_settings {
              - id    = "IntelProcVtd" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "Sriov" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "InternalSDCardSlot" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "CollabPowerControl" -> null
              - value = "Disabled" -> null
            }
          - overridden_settings {
              - id    = "UsbControl" -> null
              - value = "InternalUsbDisabled" -> null
            }
          - overridden_settings {
              - id    = "TimeFormat" -> null
              - value = "Utc" -> null
            }
          - overridden_settings {
              - id    = "ProcVirtualization" -> null
              - value = "Disabled" -> null
            }
        }

      - boot {
          - boot_order  = [
              - "HardDisk",
            ] -> null
          - manage_boot = true -> null
        }

      + boot_mode {
          + manage_mode     = true
          + mode            = "UEFI"
          + pxe_boot_policy = "Auto"
          + secure_boot     = "unmanaged"
        }

      ~ connection_settings {
            manage_connections = false

          ~ connections {
                allocated_mbps    = 0
                allocated_vfs     = 0
                function_type     = "Ethernet"
                id                = 1
                interconnect_port = 0
                isolated_trunk    = false
              - mac_type          = "Physical" -> null
              - managed           = true -> null
                maximum_mbps      = 0
                name              = "G1_Network_1"
                network_uri       = "/rest/ethernet-networks/44155074-0d29-4dcd-ba76-ec095130fc36"
                port_id           = "Mezz 3:1-a"
                requested_mbps    = "1000"

              ~ boot {
                    boot_vlan_id       = 0
                  - ethernet_boot_type = "PXE" -> null
                    priority           = "Primary"
                }
            }
          ~ connections {
                allocated_mbps    = 0
                allocated_vfs     = 0
                function_type     = "Ethernet"
                id                = 2
                interconnect_port = 0
                isolated_trunk    = false
              - mac_type          = "Physical" -> null
              - managed           = true -> null
                maximum_mbps      = 0
                name              = "G1_Network_2"
                port_id           = "Mezz 3:2-a"
                requested_mbps    = "0"

              ~ boot {
                    boot_vlan_id       = 0
                  - ethernet_boot_type = "PXE" -> null
                    priority           = "Secondary"
                }
            }
        }

      - local_storage {
          - initialize           = false -> null
          - manage_local_storage = false -> null

          - controller {
              - device_slot              = "Embedded" -> null
              - drive_write_cache        = "Unmanaged" -> null
              - import_configuration     = false -> null
              - initialize               = false -> null
              - mode                     = "Mixed" -> null
              - predictive_spare_rebuild = "Unmanaged" -> null

              - logical_drives {
                  - accelerator         = "Unmanaged" -> null
                  - bootable            = true -> null
                  - name                = "OS" -> null
                  - num_physical_drives = 2 -> null
                  - num_spare_drives    = 0 -> null
                  - raid_level          = "RAID1" -> null
                  - sas_logical_jbod_id = 0 -> null
                }
            }
        }
    }

**Comment by AsisBagga** [comment]
> Like I said the weirdness with this newer branch, was intermittent. Will continue to test as I see it was merged.
> 
> I am hoping that update_type=put won't be required in the future, as terraform should know the difference between apply changes that are "new resources" versus updating existing ones. This argument seems unintuitive. (Like how am I supposed to know (or care) what state this option should be in)
> 
> This seems silly that the end user should need to know the mechanics of the restapi call as it's an implementation detail. I didn't see that in the example either, and the doc only mentions put or patch.
> 
> Thanks for all your help! PRovider getting better everyday!

Thanks, We will add this to our backlog, and in near future you should be able to see a good documentation.

**Comment by rismoney** [comment]
what version of TF are you using with master branch?

**Comment by rismoney** [comment]
i hope the backlog isn't too long...Right now I am still really in the state of creation not updating.  I have 20+ frames coming in.

The bigger concern for me, is that right now the current release is non functional, and I am locked into a previous release with lifecycle ignores on bios and connection_settings.  Not sure what other information I can provide to you.

I think you do understand my current situation based on https://github.com/HewlettPackard/terraform-provider-oneview/issues/295#issuecomment-825686076

**Comment by AsisBagga** [comment]
I understand that you have a server profile created from a server template which had local storage set in it. 
So that means Server Profile post apply will have all the attributes which server profile template had.

Now if you trying to update server profile it is making the existing local storage set to null. 

So if you want to do an update which would need update_type="put" 
So if you do an import of the same server profile it will have all those new attributes to it which it inherited from the template. Now you will need to have the exact imported configuration in your example file, otherwise not providing the local storage in server profile example it will set the value as null because terraform knows from state file that Oneview's server profile has the local storage right now and since you are not providing them in your example means you don't need that anymore and sets to null, similar behavior is mostly with all other attributes too. 

I understand your requirement that you may not want to provide all the attributes in order to update one single attribute in the resource.

But currently we do not support this, so I would advice you to provide all the existing attributes in your example file and then try to update. 

PS: I am using Terraform v0.13.6

**Comment by rismoney** [comment]
>I understand that you have a server profile created from a server template which had local storage set in it.
>So that means Server Profile post apply will have all the attributes which server profile template had.

If I understand what you are saying is that the server profile post apply, will then not know which attributes are coming from the server profile template anymore versus what is actually in the server profile directly.  Is this correct?   This is a problematic position 
for the provider, in that in cannot ascertain "inherited" attributes.  

> Now if you trying to update server profile it is making the existing local storage set to null.
By specifying the attribute `template` I would think it would bring in the template's attributes.  That is my contention.


> So if you do an import of the same server profile it will have all those new attributes to it which it inherited from the template. 
> Now you will need to have the exact imported configuration in your example file, otherwise not providing the local storage in > server profile example it will set the value as null because terraform knows from state file that Oneview's server profile has the > local storage right now and since you are not providing them in your example means you don't need that anymore and sets to > null, similar behavior is mostly with all other attributes too.

Apparently this doesn't respect the attribute `template` to contain those settings.

I was not trying to do an update - At this point I am just running a TF plan (of non changed code) after an apply which I would expect to not have drift, and not require a code change (to add update_type).  I am just validating my existing configuration.  My belief is that something is just not right in the way terraform behavior is handling these resources declaratively, because after an apply, if no changes were made to the code base, I should end up with a 0 add, 0 changes, 0 deletes but this isn't the case.  

Maybe this is an issue with assumptions the rest api is making about the relationship between profiles and templates.

Let me see if I can break this into my actual use case, to clarify the issue:

Using TF.
1. I create a server profile template, and put a RAID 1 in it.  Apply.  No issues here.
2. I create a server profile using that template and specify a network connection in profile for ethernet vlan100 1gig. Apply. No issues.
3. Now no code changes since step 2.  I just run terraform plan now - And the server profile thinks that I want to remove my RAID1, because it's not specified in the server profile, but rather it is "inherited/merged/munged/whatever" from the serverprofiletemplate. 
4. This isn't really workable/maintainable for a TF code base....
5. ?

**Comment by rismoney** [comment]
What I am trying to articulate is that if the provider breaks terraform's intentions of defining the state and management of idempotency of server profiles this isnt workable.

If my code to create server profiles that uses templates cannot be applied/planned multiple times safely, then this resource isn't safe for usage.

It has to be a safe operation to apply the same code assuming no code changes or else something isn't evaluating or implementing properly.

**Comment by rismoney** [comment]
I am not a go developer, but I noticed this link from hashicorp: https://www.terraform.io/docs/extend/resources/customizing-differences.html

My belief is that if a server profile resource is created using a template, we may need to load the settings from the template and "assume" they're part of the profile configuration when analyzing all plans going forward.  They can't be just used for creation, and then require code changes for subsequent modification.  If a bios setting is in a template, and then a profile is created from this template, then if this relationship is not preserved, a user's code base becomes untenable.  We can't add the bios setting to the profile for the purpose of making the state clean.  It undermines templating altogether.

**Comment by AsisBagga** [comment]
@rismoney Hi Rich, I understand your requirement, despite the requirement that is not straightforward because there are other Oneview considerations, this requirement has been added to our backlog. 
The requirement is as follows:  Idempotency requirement for the server profile with template URI.

**Comment by nabhajit-ray** [comment]
Hi @rismoney ,
The issue is root caused the same is getting reviewed . We are planning release the fix with OV 6.2 release

**Comment by rismoney** [comment]
Just figured I would provide some updates.

I was having some weirdness on network connections, which I may or may not have isolated - TBD.  To create network connections I use a data "oneview_ethernet_network".   If I were to pass an empty "name" into this, I may have been getting erratic results.  I now did this `network_uri = connections.value.uri != "" ? connections.value.uri : ""` for the connection_settings connections, and that may have fixed that.  Still testing.   I also had an odd case, where the network connection it the data network picked was not even defined in the code.  Haven't been able to reproduce again, but will keep you posted.  

We have use cases where we have network connections that are intentionally unassigned, and believe that may have been causing the strange behavior on the network uri.

On the topic of profile name matching - on latest master branch, I still occasionally see mismatching of servers.  Note, I destroyed all server profiles and redeployed everything on this version.

Example: [snipped]
```
      id                    = "SP_EG01_FRAME01_B01"
      ~ name                  = "SP_EG01_FRAME02_B05" -> "SP_EG01_FRAME01_B01"

```
Not sure where it randomly gets Frame2 Bay 05 from, but, usually a subsequent terraform plan resolves this correctly so something is definitely wonky on that front.  It usually happens seemingly random after about 25-30 profiles have been deployed through modules.  It is correct in state file. Using TF 0.12.30 so I don't know if it is something with modules that may be fixed in new release or the provider.  But it is definitely an oddball scenario I can't explain based on my TF files.
